lib/services.js

```diff
@@ -191,6 +201,29 @@ export default {
                  });
     },
 
+   /**
+    * Check that hashedStream.completedHash matches header contentMd5.
+    * @param {object} request - request linked to the stream containine the data
+    * @param {string} hashedStream - hased stream once completed
+    * @param {object} dataRetrievalInfo - dataStore: backend stored key
+    * {key, dataStoreName (mem/file)}
+    * @param {RequestLogger} log - the current request logger
+    * @return {object} errors
+    */
+    checkHashMatchMD5(request, hashedStream, dataRetrievalInfo, log) {
+        if (request.contentMD5 && hashedStream.completedHash &&
+        request.contentMD5 !== hashedStream.completedHash) {
+            // delete data if does not match.
+            data.batchDelete(dataRetrievalInfo, log);
```

Comment:
This should go in the method `data.put` where `checkHashMatchMD5` is being called for separation of concerns.